Initial Implementation - v3

andreeaflorescu opened the pull request:
Added an initial implementation of vmm-epoll. This is loosely based on the implementation we have in Firecracker (at the interface level), but goes further into fixing some loose ends described in this comment: https://github.com/rust-vmm/vmm-epoll/pull/2#issuecomment-586204636

jiangliu commented on src/event_manager.rs:
```diff
@@ -0,0 +1,493 @@
+
+/// A trait to express the ability to respond to I/O event readiness
+/// using callbacks.
+pub trait Subscriber {
```
how about change to "pub trait Subscriber: Send"? so we run multiple IO worker threads to handle epoll events.

jiangliu replied:
Could we add some writer/maintainer to the vmm-epoll repo so we could approve the PR?

alexandruag replied:
I know this is something that comes up often, but is it very inconvenient if we do not add that restriction to the trait itself, but rather from the outside when using it? Adding the bound here means all subscribers will have to be `Send` which might not fit all potential use cases.

Also, how do you envision running the multiple IO worker threads? Will there be a separate event manager on every thread, or do you prefer some other model?

jiangliu commented:
I like this new design:) 
It uses a hashmap instead of a vector to store handler, and gets rid of the requirement to allocate an ID for each handler. We have switched to Hashmap too, but the legacy handler ID is still kept for compatible. If firecracker is migrating off the old vector based design, we are happy to adopt a hashmap based design.
Thanks for moving away from `Vector`, it makes things complex.

jiangliu commented:
Could we add some writers/maintainers to the vmm-epoll repo so we could approve the PR?

andreeaflorescu commented:
@jiangliu I added the gatekeepers to the repository. You should have write permissions now. The PR is not really ready to merge, I have to make some improvements. I will create a TODO list and keep you post it.

jiangliu commented:
> @jiangliu I added the gatekeepers to the repository. You should have write permissions now. The PR is not really ready to merge, I have to make some improvements. I will create a TODO list and keep you post it.

Thanks, @andreeaflorescu .
Any update about this PR? I want to give a try to it:)

andreeaflorescu commented:
> Thanks, @andreeaflorescu .
> Any update about this PR? I want to give a try to it:)
@jiangliu I still don't have the best solution in mind right now, this is more of a draft. Do you want to merge this and do incremental changes on top of it? I think the interface might change a lot with the changes that we need to do.

So far, this code has the following problems:
1. We do not have any clean up for the epoll fd. This is rather easy to fix and I already have a PR for this in Firecracker: https://github.com/firecracker-microvm/firecracker/pull/1592
2. We don't restrict the operations that the `Subscriber` can call on the EventManger. A try to fix this is here: https://github.com/firecracker-microvm/firecracker/pull/1587, but I am not very happy with how things look at the interface level.
3. We have 2 functions that mostly do the same: `add_subscriber` and `register`. Ideally we should only use `register` and get rid of `add_subscriber`. This is not an easy task though and to fix it we might need to introduce a channel or have an activate event of some sorts. This is a bit weird, and we should discuss about this at length.
4. We are working with `RawFd`s. (yacx). This is not particularly safe because the objects from which those RawFds were created might not actually be alive when we are using them.
5. We kinda pretend to give full control to the Subscriber in terms of what the registration data in `epoll_event` is, but actually if the Subscriber doesn't put a `RawFd` there, the EventManager doesn't really work. This should be fixed as well. We can create yet another wrapper over `EpollEvent` and only allow data to be something that implements `AsRawFd`. A better option though is to segment the `data` field of the `EpollEvent` and give the user the possibility to set 32 bits with whatever data they please. We would still need yet another wrapper over `EpollEvent` that is safe to use.
6. The recently discovered problem with INTR: https://github.com/firecracker-microvm/firecracker/issues/1604 
7. There might be other things as well that I am missing now.

Points 3, 4, and 5 need a separate design discussion. We can have that in a separate issue in vmm-epoll.

Most of these can be parralelized, so multiple people can work on having this crate in a good shape faster. Is this something that we would like to do? If so, we could merge this PR, but point out that it is still WIP and the interface might severely change. What do you think?

CC: @alexandruag @sameo

alexandruag commented:
Hi! Like Andreea mentioned, there is quite the list of things we'd like to fix first, and right now it looks like the end of next week is when we'll have either solutions or a much better idea about the path forward. Does that sound too late?

jiangliu commented:
> Hi! Like Andreea mentioned, there is quite the list of things we'd like to fix first, and right now it looks like the end of next week is when we'll have either solutions or a much better idea about the path forward. Does that sound too late?

Thanks, @andreeaflorescu  and @alexandruag 
Looking forwarding for the new implementation:)

andreeaflorescu commented:
@jiangliu we need a couple of more days. We are wrapping up the interface definition and adding tests & documentation. Sorry for the delay.

jiangliu commented:
Thanks, Andreea:)
What's the plan for firecracker to switch to the new vmm-epoll?

andreeaflorescu commented:
We are now merging this version into Firecracker master because it was developed in a separate branch: https://github.com/firecracker-microvm/firecracker/pull/1618

The plan is to do the necessary interface change and updates directly in rust-vmm and then consume vmm-epoll from here.

andreeaflorescu commented:
Hey @jiangliu. @alexandruag and I worked on fixing the issues with the initial implementation of the EventManager specified in my comment https://github.com/rust-vmm/vmm-epoll/pull/2#issuecomment-586204636.

At the interface level, the `EventSubscriber` and `EventManager` should be pretty similar. We updated the implementation to fix the underlying problems. Alex also added a channel to make it easy to communicate with the `EventManager` from a different thread.

It's quite a lot of code now, please let us know if you have any questions. Also, sorry for the delay, we got a bit side tracked by other stuff.

acatangiu commented on src/event_manager/manager.rs:
```diff
@@ -0,0 +1,415 @@
+};
+
+/// Allows event subscribers to be registered, connected to the event loop, and later removed.
+pub struct EventManager<T> {
+    subscribers: Subscribers<T>,
+    epoll_context: EpollContext,
+    ready_events: Vec<EpollEvent>,
+    channel: EventManagerChannel<T>,
+}
```
I'm guessing that this struct is made of generic type `T` so that the users can either do:
- `EventManager<Arc<Mutex<dyn EventSubscriber>>>` - in a multithreaded model
  or
- `EventManager<Rc<RefCell<dyn EventSubscriber>>>`- in a singlethreaded model

right?

I know we can also directly do `EventManager<X: EventSubscriber>`, but that means the `EventManager` takes exclusive ownership of `X` so I don't think this use-case is actually useful.

### Question:
Would it make sense to drop the generic `T` and provide two external interfaces instead of one to the EventManager, one multi-thread-safe `Arc<Mutex<dyn EventSubscriber>>` and one single-thread `Rc<RefCell<dyn EventSubscriber>>`?

**Pros:**
- A single `EventManager` can handle both single-thread and multi-thread objects at the same time. The current generic can't, it's either one or the other.
- The code should become much simpler without the generics.

**Cons:**
- We lose the case `EventManager<X: EventSubscriber>`, but like I said, I'm not sure how useful that is in practice anyway.
- More code, two paths, single and multi-threaded, but should be easy to follow.

What do you think?

jiangliu replied:
It will be great if we could remove the generic parameter, there's already a wrapper:
```
impl EventSubscriber for Arc<Mutex<dyn EventSubscriber>>
```

andreeaflorescu replied:
@jiangliu What is your concern with using the generic parameter?

I personally don't think we should restrict the interface if we don't have any gain from it. Using `Arc<Mutex<..>>` is not necessary in all VMMs and it isn't a zero cost abstraction. Besides the performance hit that you might get, it is also pretty annoying to use Arc<Mutext<>> because you have to write `.lock().unwrap()` with each usage of the `EventSubscriber`.

In any case, I will get back to you on this topic after I sync with @alexandruag as well.

acatangiu commented:
I'm proposing some small changes in https://github.com/andreeaflorescu/vmm-epoll/pull/1 to simplify part of the code. What do you think?

alexandruag commented:
Thanks for the suggestions! All these have sparked quite a lot of questions, and I think the best way to proceed is to step back and think about the use cases we're trying to enable with this event manager proposal. I've started writing an outline along those lines, and hopefully we can use that and any other ideas ppl have to better put the implementation and trade-offs into perspective.

andreeaflorescu commented:
> I'm proposing some small changes in [andreeaflorescu#1](https://github.com/andreeaflorescu/vmm-epoll/pull/1) to simplify part of the code. What do you think?

@acatangiu I've incorporated most of the changes you suggested (I think) in the current PR. Can you please take another look?

acatangiu commented on src/event_manager/control_ops.rs:
```diff
@@ -0,0 +1,79 @@
+    // Mutable reference to the EpollContext of an EventManager.
+    epoll_context: &'a mut EpollContext,
+    // The index (within the subscribers member of the EventManager) of the event subscriber
+    // this object stands for.
```
```suggestion
    // The unique id of the event subscriber this object stands for.
```

acatangiu commented on src/event_manager/epoll_context.rs:
```diff
@@ -0,0 +1,57 @@
+pub(crate) struct EpollContext {
+    // The epoll wrapper.
+    pub(crate) epoll: Epoll,
+    // Records the id of the subscriber associated with the given RawFd. The event event_manager
```
```suggestion
    // Records the unique id of the subscriber associated with the given RawFd. The event_manager
```

acatangiu commented on src/event_manager/epoll_context.rs:
```diff
@@ -0,0 +1,57 @@
+    // does not currently support more than one subscriber being associated with an fd.
+    pub(crate) fd_dispatch: HashMap<RawFd, SubscriberId>,
+    // Records the set of fds that are associated with the subscriber that has the given index
+    // (this is essentially the reverse of the previous field).
```
```suggestion
    // Records the set of fds that are associated with the subscriber that has the given unique id
    // (this is used to keep track of all fds registered to a specific subscriber).
```

acatangiu commented on src/event_manager/epoll_context.rs:
```diff
@@ -0,0 +1,57 @@
+
+    // Remove the fds associated with the provided subscriber id from the epoll set and the
+    // other structures. The subscriber id must be valid.
+    pub(crate) fn remove(&mut self, subscriber_index: SubscriberId) {
```
```suggestion
    pub(crate) fn remove(&mut self, subscriber_id: SubscriberId) {
```

acatangiu commented on src/event_manager/epoll_context.rs:
```diff
@@ -0,0 +1,57 @@
+    pub(crate) fn remove(&mut self, subscriber_index: SubscriberId) {
+        let fds = self
+            .subscriber_watch_list
+            .remove(&subscriber_index)
```
```suggestion
            .remove(&subscriber_id)
```

acatangiu commented on src/event_manager/epoll_context.rs:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    // Gets the id of the subscriber associated with the provided fd (if such an association
+    // exists).
```
```suggestion
    // Gets the unique id of the subscriber associated with the provided fd
    // (if such an association exists).
```

acatangiu commented on src/event_manager/subscribers.rs:
```diff
@@ -0,0 +1,51 @@
+    //
+    // This method should only be called for indices that are known to be valid, otherwise
+    // panics can occur.
+    pub(crate) fn get_mut(&mut self, subscriber_id: SubscriberId) -> &mut T {
```
Since this does `unwrap` without any checks, I suggest we change its name to `fn get_mut_unchecked()`.

```suggestion
    pub(crate) fn get_mut_unchecked(&mut self, subscriber_id: SubscriberId) -> &mut T {
```

andreeaflorescu commented:
@jiangliu can you take another look? I would like to add tests and documentation, but I would do that after we agree with the general design of the crate. What do you say?

alexandruag commented:
Hi, the code looks nicer now! Going back to the discussion about dropping the generic `T`, I think that loses some flexibility along three dimensions: users might opt for static instead of dynamic dispatch, subscribers that implement interior mutability won't need things like `Arc`/`Rc`, and some use cases might rely on custom wrappers.

One thing that's still a bit clunky now (compared to, for example, `polly` from Firecracker) is the two step "get id from hashmap" -> "get subscriber from hashmap" resolution. A way around that is to parametrize the manager with a reference counted subscriber container type as well (or have that as the only type parameter), that can be then used to dispatch events in one step (by adding clones of the container in the `fd_dispatch` hashmap directly). Types such as `Arc<Mutex<...>>` can be their own container, so there's no extra wrapping. Is anyone interested in seeing a prototype of this change?

I was also wondering if the `RemoteEndpoint` abstraction makes sense, and if it meets any existing use cases. I think it's helpful for interacting with the manager from different threads, without the need for more intrusive synchronization. One example is when calling `EventManager` methods following events that happen asynchronously with respect to its operation (e.g. vcpu exit events). Another example use case is getting access to subscribers that are moved to different event manager threads for event handling, but are not wrapped in a synchronized handle like `Arc<Mutex<...>>`. We should consider putting the `RemoteEndpoint` functionality behind a feature, or removing it entirely for the time being if it doesn't really look useful.

alexandruag commented:
I've played around a bit more and turns out changing the dispatch approach to the one mentioned in my pervious comment has some implications regarding the interface as well. Overall, I think the current iteration is in a pretty good spot; we can think about fine tuning it after we get more input/data.

jiangliu commented on src/event_manager/subscribers.rs:
```diff
@@ -0,0 +1,51 @@
+    // We are generating the unique ids by incrementing this counter for each added subscriber,
+    // and rely on the extremely large value range of u128 to ensure each value is effectively
+    // unique over the runtime of any VMM.
+    next_id: u128,
```
Why u128 is preferred instead of u64? I think u64 should be bigger enough and more efficient than u128.

andreeaflorescu replied:
I'm fine with using u64 instead, @alexandruag what do you think?

sboeuf replied:
Yes let's use u64.

jiangliu commented on src/event_manager/subscribers.rs:
```diff
@@ -0,0 +1,51 @@
+    pub(crate) fn new() -> Self {
+        Subscribers {
+            subscribers: HashMap::new(),
+            next_id: 0,
```
How about reserving subscriber_id 0?

andreeaflorescu replied:
Do you see a problem with starting from 0?

jiangliu replied:
No really problem, just want to reserve id 0 for the RemoteEndpoint fd.
And reserving id 0 may help to debug in some cases:)

andreeaflorescu replied:
I'll update it.

jiangliu commented on src/event_manager/manager.rs:
```diff
@@ -0,0 +1,416 @@
+pub struct EventManager<T> {
+    subscribers: Subscribers<T>,
+    epoll_context: EpollContext,
+    ready_events: Vec<EpollEvent>,
```
Like the idea to use a statically  allocated buffer for EpollEvemt, actually we have adopted a similar mechanism to avoid allocating/freeing memory on each epoll loop.

andreeaflorescu commented on src/epoll.rs:
```diff
@@ -0,0 +1,433 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Will do that after we publish a new version of vmm-sys-util. The first commit here: https://github.com/rust-vmm/vmm-sys-util/pull/83 adds some derives that are needed in this PR.

alexandruag replied:
Hmm, if the `epoll` wrapper ends up residing in `vmm-sys-util` and only the manager remains here, then the current name of the crate doesn't seem such a good fit anymore :( WDYT about this aspect?

jiangliu replied:
The first implementation from me uses the epoll crate and the vmm-epoll only contains the manager:)

andreeaflorescu commented on src/event_manager/mod.rs:
```diff
@@ -0,0 +1,58 @@
+    /// Register the events initially associated with the subscriber.
+    ///
+    /// Called by the `EventManager` after a subscriber is registered.
+    fn init(&self, ops: &mut ControlOps);
```
> how about taking a" &mut self"?
Hmmm, I am not sure you need to mutate self. Do you? If that's the case, we can change it to mut.

`ControlOps` was intentionally added as an abstraction here to fix point 2 in the TODO list: https://github.com/rust-vmm/vmm-epoll/pull/2#issuecomment-586204636.

The idea behind `ControlOps` was to hide the fact that the event manager works with `epoll` and also limit the operations that you can do from the subscriber. This way we make sure that only valid operations are called: `add`, `remove`, `modify`.

jiangliu replied:
> > how about taking a" &mut self"?
> 
> Hmmm, I am not sure you need to mutate self. Do you? If that's the case, we can change it to mut.
> 
> `ControlOps` was intentionally added as an abstraction here to fix point 2 in the TODO list: [#2 (comment)](https://github.com/rust-vmm/vmm-epoll/pull/2#issuecomment-586204636).
> 
> The idea behind `ControlOps` was to hide the fact that the event manager works with `epoll` and also limit the operations that you can do from the subscriber. This way we make sure that only valid operations are called: `add`, `remove`, `modify`.

We could achieve the goal of [#2](https://github.com/rust-vmm/vmm-epoll/pull/2#issuecomment-586204636) by
* add argument "subscriber_id" to EventSubscriber::init()
* make EventSubscriber::init() takes &mut self
* export EpollContext directly
* move add/modify/remove from ControlOps to EpollContext and make them pub.
By this way, it gets simpler and achieves a minor performance improve.

andreeaflorescu replied:
I don't like the idea of exporting `EpollContext` with its current form and name because it `Context` doesn't really mean anything and `Epoll` is strongly suggesting that everything is based on `epoll` :)). I will try to find an abstraction that makes more sense at the interface level, while keeping in mine your feedback. To Alex's comment https://github.com/rust-vmm/vmm-epoll/pull/2#discussion_r402621389, what do you think about renaming this crate to `event-manager` instead? This way we could leave it open for other implementations as well. The interface could be extended to work on Windows as well or support other "backends". WDYT?

jiangliu replied:
event-manager is OK to me, to be more agnostic.

andreeaflorescu replied:
Cool. Event Manager seems to be available on crates.io as well.

andreeaflorescu replied:
@jiangliu with this approach, you would need to check the validity of the <SubscriberId, Events> pair when calling process. I didn't do any benchmarks, but I assume that the minor performance improvement will not be visible anymore because of the extra checks that are needed.

jiangliu replied:
For most epoll events, the handler will do read/write operations instead of EventOperations::{add, remove, modify}.

andreeaflorescu replied:
Hmm, do you mean that during process you are mostly just processing events and not using the epoll reference to alter the events?

sboeuf commented on src/event_manager/control_ops.rs:
```diff
@@ -0,0 +1,78 @@
+/// `EventManager`.
+// Right now this is a concrete object, but going further it can be turned into a trait and
+// passed around as a trait object.
+pub struct ControlOps<'a> {
```
Why don't we simply make the association between the EpollContext and the subscriber ID part of EventManager directly? That would remove the indirection layer and the opaque object.

alexandruag replied:
Can you please provide a very short example here? (for example, how do you see the signature of any method impacted by your suggestion changing with the removal of the indirection layer). I have a couple of different things in mind that kinda fit your description, but I'm not quite sure which one actually matches.

andreeaflorescu commented on src/event_manager/epoll_context.rs:
```diff
@@ -0,0 +1,57 @@
+    pub(crate) epoll: Epoll,
+    // Records the id of the subscriber associated with the given RawFd. The event event_manager
+    // does not currently support more than one subscriber being associated with an fd.
+    pub(crate) fd_dispatch: HashMap<RawFd, SubscriberId>,
```
The key in the hashmap is the `RawFd`. When you want to add map the same fd with another subscriber ID in the `fd_dispatch`, the previous one will be overridden.

alexandruag commented on src/event_manager/endpoint.rs:
```diff
@@ -0,0 +1,136 @@
+
+// Until more sophisticated methods are explored (for example making the `EventManager` offer
+// interior mutability using something like an RCU mechanism), the current proposal relies on
+// passing boxed closures to the manager and getting back a boxed result. The manager is notified
```
Hmm, the signature of the closure (`FnOnce(..) -> ...`) is a trait itself, and that's why I was under the impression this is not much different from using any other trait object within a `Box`. Are you referring to some other level of indirection?

alexandruag commented on src/event_manager/events.rs:
```diff
@@ -0,0 +1,75 @@
+/// keep the ability to associate opaque data with an event. An `Events` object always contains an
+/// fd and an `u32` data member that can be supplied by the user. When registering events with the
+/// inner epoll event set, the fd and data members of `Events` as used together to generate the
+/// underlying `u64` member of the epoll_data union.
```
From my (limited, to be fair) previous experience, using `EpollEvent` as the abstraction here gets quite tricky if we want to allow users to associate a token with an event as well.

The standard `epoll` interface (which is accurately reflected by `EpollEvent` as our low-level wrapper) expects the user to associate an opaque 64bit value with each registered event, and this value gets returned by `epoll_wait` to identify the source of the event. Presently, the event manager relies on fds (32 bit wide) to uniquely identify the source of every event, but we also wanted to allow users to set/receive the remaining 32 bits as opaque data. The `Events` abstraction transparently handles this, as we've previously encountered bugs when doing the shifting and/or reinterpretation by hand in `EpollEvent`.

andreeaflorescu commented:
I tried to merge `EpollContext` with `ControlOps` and this is the least ugly solution from what I tried so far: https://github.com/andreeaflorescu/vmm-epoll/commit/722acd4f3c0df5abd3fca922e073e4f75ad92d57.

@sboeuf @jiangliu @alexandruag do you have other proposals for merging the two?

If we can't find something better, I would propose to keep them as they are and move stuff around, name them better, provide documentation. I find the user interface to be more important in this case and I wouldn't want to merge `EpollContext` with `ControlOps` at the expense of the UI. WDYT?

andreeaflorescu commented on src/event_manager/endpoint.rs:
```diff
@@ -0,0 +1,136 @@
+
+use super::{Error, EventSubscriber, Result, SubscriberOps};
+
+// TODO: gate the remote endpoint functionality behind a feature?
```
Done. I created the remote-endpoint.

jiangliu commented:
> I tried to merge `EpollContext` with `ControlOps` and this is the least ugly solution from what I tried so far: [andreeaflorescu@722acd4](https://github.com/andreeaflorescu/vmm-epoll/commit/722acd4f3c0df5abd3fca922e073e4f75ad92d57).
> 
> @sboeuf @jiangliu @alexandruag do you have other proposals for merging the two?
> 
> If we can't find something better, I would propose to keep them as they are and move stuff around, name them better, provide documentation. I find the user interface to be more important in this case and I wouldn't want to merge `EpollContext` with `ControlOps` at the expense of the UI. WDYT?

After reading the code again, I'm hesitating whether we need the Subscribers/SubscriberId at all.
What happens if we use the rawfd as the SubscriberId?
I will do some experiment based on https://github.com/andreeaflorescu/vmm-epoll/tree/event_manager_v4

acatangiu commented:
> After reading the code again, I'm hesitating whether we need the Subscribers/SubscriberId at all.
What happens if we use the rawfd as the SubscriberId?
I will do some experiment based on https://github.com/andreeaflorescu/vmm-epoll/tree/event_manager_v4

I'd advise against it. `RawFd`s are not unique identifiers, they can be reused at the process level. If the code using `EventManager` releases the resource, the fd number is up for grabs. Even if `epoll` documentation says `epoll` will remove the fd on release, the `EventManager` is not implicitly made aware of the event and can end up with inconsistent internal state. Trying to add a new resource that happens to have reused the previous fd#, will result in UB as opposed to being correctly supported with the `SubscriberId` version.

jiangliu commented:
> > After reading the code again, I'm hesitating whether we need the Subscribers/SubscriberId at all.
> > What happens if we use the rawfd as the SubscriberId?
> > I will do some experiment based on https://github.com/andreeaflorescu/vmm-epoll/tree/event_manager_v4
> 
> I'd advise against it. `RawFd`s are not unique identifiers, they can be reused at the process level. If the code using `EventManager` releases the resource, the fd number is up for grabs. Even if `epoll` documentation says `epoll` will remove the fd on release, the `EventManager` is not implicitly made aware of the event and can end up with inconsistent internal state. Trying to add a new resource that happens to have reused the previous fd#, will result in UB as opposed to being correctly supported with the `SubscriberId` version.

Per my understanding, current design can't protect us from the issue too:)
Though we allocate SubscribeId for each subscriber, different SubscriberIds  may still map to the same RawFd, though cause inconsistent internal state. Let's image a sequence:
1) EventManager thread calls run_with_timeout()
    ->dispatch_events()
           ->handle_endpoint_calls()
                 ->remove_subscriber() for rawfd1
                 ->................
                 ->add_subscriber() for rawfd1 (the rawfd1 get reused after it has been closed)
           -> handle epoll event triggered by the original rawfd1
                the subscriber registered for the new/reused rawfd1 gets invoked (get wrong here)

acatangiu commented:
> Per my understanding, current design can't protect us from the issue too:)
Though we allocate SubscribeId for each subscriber, different SubscriberIds may still map to the same RawFd, though cause inconsistent internal state. Let's image a sequence:

    EventManager thread calls run_with_timeout()
    ->dispatch_events()
    ->handle_endpoint_calls()
    ->remove_subscriber() for rawfd1
    ->................
    ->add_subscriber() for rawfd1 (the rawfd1 get reused after it has been closed)
    -> handle epoll event triggered by the original rawfd1
    the subscriber registered for the new/reused rawfd1 gets invoked (get wrong here)

From the [epoll man page](http://man7.org/linux/man-pages/man7/epoll.7.html):
```
Q6:
           Will closing a file descriptor cause it to be removed from all
           epoll interest lists?

           Yes, but be aware of the following point.  A file descriptor is a
           reference to an open file description (see open(2)).  Whenever a
           file descriptor is duplicated via dup(2), dup2(2), fcntl(2)
           F_DUPFD, or fork(2), a new file descriptor referring to the same
           open file description is created.  An open file description con‐
           tinues to exist until all file descriptors referring to it have
           been closed.

           A file descriptor is removed from an interest list only after all
           the file descriptors referring to the underlying open file
           description have been closed.  This means that even after a file
           descriptor that is part of an interest list has been closed,
           events may be reported for that file descriptor if other file
           descriptors referring to the same underlying file description
           remain open.  To prevent this happening, the file descriptor must
           be explicitly removed from the interest list (using epoll_ctl(2)
           EPOLL_CTL_DEL) before it is duplicated.  Alternatively, the
           application must ensure that all file descriptors are closed
           (which may be difficult if file descriptors were duplicated
           behind the scenes by library functions that used dup(2) or
           fork(2)).
```

So as long as the resource is fully dropped (all fds referencing it are closed), `epoll` automatically removes the fd from its interest list. This case is nicely handled with `SubscriberId` and would not be with `RawFd` as identifier.

In case of `dup`-in FDs and closing the one added to `EventManager` we end up with UB regardless of whether we're using `SubscriberId` or `RawFd`.
I guess at the very least this should be properly documented.

andreeaflorescu commented:
@jiangliu As Adi explained, we choose this design where the SubscriberId is generated by the `EventManager` and is truly unique at the process level (i.e. is not a RawFd that can be re-assigned) to counter the problem you are describing.

With `SubscriberId` as an u32, we should not run into that problem because the subscriber will receive a hang up error and the subscriber can handle that error accordingly.

jiangliu commented:
> > Per my understanding, current design can't protect us from the issue too:)
> > Though we allocate SubscribeId for each subscriber, different SubscriberIds may still map to the same RawFd, though cause inconsistent internal state. Let's image a sequence:
> 
> ```
> EventManager thread calls run_with_timeout()
> ->dispatch_events()
> ->handle_endpoint_calls()
> ->remove_subscriber() for rawfd1
> ->................
> ->add_subscriber() for rawfd1 (the rawfd1 get reused after it has been closed)
> -> handle epoll event triggered by the original rawfd1
> the subscriber registered for the new/reused rawfd1 gets invoked (get wrong here)
> ```
> 
> From the [epoll man page](http://man7.org/linux/man-pages/man7/epoll.7.html):
> 
> ```
> Q6:
>            Will closing a file descriptor cause it to be removed from all
>            epoll interest lists?
> 
>            Yes, but be aware of the following point.  A file descriptor is a
>            reference to an open file description (see open(2)).  Whenever a
>            file descriptor is duplicated via dup(2), dup2(2), fcntl(2)
>            F_DUPFD, or fork(2), a new file descriptor referring to the same
>            open file description is created.  An open file description con‐
>            tinues to exist until all file descriptors referring to it have
>            been closed.
> 
>            A file descriptor is removed from an interest list only after all
>            the file descriptors referring to the underlying open file
>            description have been closed.  This means that even after a file
>            descriptor that is part of an interest list has been closed,
>            events may be reported for that file descriptor if other file
>            descriptors referring to the same underlying file description
>            remain open.  To prevent this happening, the file descriptor must
>            be explicitly removed from the interest list (using epoll_ctl(2)
>            EPOLL_CTL_DEL) before it is duplicated.  Alternatively, the
>            application must ensure that all file descriptors are closed
>            (which may be difficult if file descriptors were duplicated
>            behind the scenes by library functions that used dup(2) or
>            fork(2)).
> ```
> 
> So as long as the resource is fully dropped (all fds referencing it are closed), `epoll` automatically removes the fd from its interest list. This case is nicely handled with `SubscriberId` and would not be with `RawFd` as identifier.
> 
> In case of `dup`-in FDs and closing the one added to `EventManager` we end up with UB regardless of whether we're using `SubscriberId` or `RawFd`.
> I guess at the very least this should be properly documented.

A special case is that the fd to be removed is already in the self.ready_events buffer.
1) epoll.wait() returns both the channel.event_fd and another fd(say fdA). And the channel.event_fd is ahead of fdA.
2) when handling event for the channel.event_fd, a race window causes that the fdA is removed and then re-inserted.
3) when handling the pending event for legacy fdA in the ready_events buffer, the newly registered subscriber will be invoked.

To solve this problem, we may change handle_endpoint_calls() to only handle one event per loop, thus breaks the race window in step 2. Once the race window has been removed, the indirect mapping from rawfd number to subcriberid seems redundant.

jiangliu commented:
> @jiangliu As Adi explained, we choose this design where the SubscriberId is generated by the `EventManager` and is truly unique at the process level (i.e. is not a RawFd that can be re-assigned) to counter the problem you are describing.
> 
> With `SubscriberId` as an u32, we should not run into that problem because the subscriber will receive a hang up error and the subscriber can handle that error accordingly.

The SubscriberId is unique, but the mapping chain "rawfd -> SubscriberId-> subscriber" may not be unique when handling pending epoll events in the "ready_events" buffer. I feel there's a race window:)

andreeaflorescu commented:
The `RawFd` -> `SubscriberId` should also be unique because we are only adding something in the hasmap if the `RawFd` does not exist. Otherwise we return an error.

I will double check though.

jiangliu commented:
> The `RawFd` -> `SubscriberId` should also be unique because we are only adding something in the hasmap if the `RawFd` does not exist. Otherwise we return an error.
> 
> I will double check though.

Please pay attention to the race window caused by pending epoll events in the event buffer:) 
And a rawfd number may get removed and re-inserted in a short period.

andreeaflorescu commented:
@jiangliu you're right! I am not though convinced that we will get a performance improvement with getting rid of all the other abstractions (EpollContext, ControlOps, SubscriberId, Subscriers). We can try to do some benchmarks after you also have a proposal here.

jiangliu commented:
> @jiangliu you're right! I am not though convinced that we will get a performance improvement with getting rid of all the other abstractions (EpollContext, ControlOps, SubscriberId, Subscriers). We can try to do some benchmarks after you also have a proposal here.

Thanks, Andreea. So let's keep current implementation and move forward:)

andreeaflorescu commented:
Thanks @jiangliu! I'll fix the build so we can merge this. I have a few extra patches on top of this one, but it is probably better to have them as subsequent PRs.
